k8sutil/kube_config.py

```diff
@@ -0,0 +1,160 @@
+
+from oauth2client.client import GoogleCredentials
+
+from k8sclient import configuration
+
+_temp_files = []
+
+
+def _cleanup_temp_files():
+    for f in _temp_files:
+        os.remove(f)
+
+
+def _create_temp_file_with_content(content):
+    if len(_temp_files) == 0:
+        atexit.register(_cleanup_temp_files)
+    _, name = tempfile.mkstemp()
+    _temp_files.append(name)
+    if isinstance(content, str):
+        content=content.encode('utf8')
+    with open(name, 'wb') as fd:
+        fd.write(base64.decodestring(content))
+    return name
+
+
+def _find_object_with_name(o, name, o_name):
+    for c in o:
+        if c['name'] == name:
+            return c
+    raise Exception("Cannot find object with name %s in %s" % (name, o_name))
+
+
+def _file_from_file_or_data(o, file_key_name, data_key_name=None):
+    if not data_key_name:
+        data_key_name = file_key_name + "-data"
+    if data_key_name in o:
+        return _create_temp_file_with_content(o[data_key_name])
+    if file_key_name in o:
+        return o[file_key_name]
+    return None
+
+
+def _data_from_file_or_data(o, file_key_name, data_key_name=None):
+    if not data_key_name:
+        data_key_name = file_key_name + "_data"
+    if data_key_name in o:
+        return o[data_key_name]
+    if file_key_name in o:
+        with open(o[file_key_name], 'r') as f:
+            data = f.read()
+        return data
+    return None
+
+
+def _load_gcp_token(user):
+    if 'auth-provider' not in user:
+        return None
+    if 'name' not in user['auth-provider'] or user['auth-provider']['name'] != 'gcp':
+        return None
+    # Ignore configs in auth-provider and rely on GoogleCredentials caching and refresh mechanism.
+    # TODO: support gcp command based token ("cmd-path" config).
+    return GoogleCredentials.get_application_default().get_access_token().access_token
+
+
+def _load_authentication(user):
+    """Read authentication from kube-config user section.
+
+    This function goes through various authetication methods in user section of kubeconfig and
+    stops if it founds a valid authentication method. The order of authentication methods is:
+
+        1. GCP auth-provider
+        2. token_data
+        3. token field (point to a token file)
+        4. username/password
+    """
+    # Read authentication
+    token = _load_gcp_token(user)
+    if not token:
+        token = _data_from_file_or_data(user, 'tokenFile', 'token')
+    if token:
+        configuration.api_key['authorization'] = "bearer " + token
+    else:
+        if 'username' in user and 'password' in user:
+            configuration.api_key['authorization'] = urllib3.util.make_headers(
+                basic_auth=user['username'] + ':' + user['password']).get('authorization')
+
+
+def _load_cluster_info(cluster, user):
+    """Loads cluster information from kubeconfig such as host and SSL certificates."""
+    if 'server' in cluster:
+        configuration.host = cluster['server']
```

Comment:
Hmm, this is a global?  Will this library be capable of contacting multiple kube clusters at the same time, or would that require multiple processes?